cubi_tk/irods/check.py

```diff
@@ -1,73 +1,100 @@
 
+        #: Path to iRODS environment file
+        self.irods_env_path = os.path.join(
+            os.path.expanduser("~"), ".irods", "irods_environment.json"
+        )
+
+        #: iRODS sessions for parallel execution
+        self.irods_sessions = []
+
+        #: iRODS environment
+        self.irods_env = None
+
+    def __del__(self):
+        # Ensure cleanup of iRODS sessions
+        for irods in self.irods_sessions:
+            irods.cleanup()
+
     @classmethod
     def setup_argparse(cls, parser: argparse.ArgumentParser) -> None:
         parser.add_argument(
             "--hidden-cmd", dest="irods_cmd", default=cls.run, help=argparse.SUPPRESS
         )
 
         parser.add_argument(
+            "-r",
             "--num-replicas",
+            dest="req_num_reps",
             type=int,
             default=MIN_NUM_REPLICAS,
             help="Minimum number of replicas, defaults to %s" % MIN_NUM_REPLICAS,
         )
 
         parser.add_argument(
+            "-n",
             "--num-parallel-tests",
             type=int,
             default=NUM_PARALLEL_TESTS,
             help="Number of parallel tests, defaults to %s" % NUM_PARALLEL_TESTS,
         )
 
-        parser.add_argument("irods_path", help="Path to an iRods collection.")
+        parser.add_argument("irods_path", help="Path to an iRODS collection.")
+
+    @classmethod
+    def get_irods_error(cls, e: Exception):
+        """Return logger friendly iRODS exception."""
+        es = str(e)
+        return es if es != "None" else e.__class__.__name__
 
-    def get_files(self):
-        """get files on iRods."""
+    def connect(self):
+        """Connect to iRODS."""
         try:
-            ils_out = check_output(["ils", "-r", self.args.irods_path]).decode(sys.stdout.encoding)
-        except SubprocessError as e:  # pragma: nocover
-            logger.error("Something went wrong: %s\nAre you logged in? try 'iinit'", e)
+            return iRODSSession(irods_env_file=self.irods_env_path)
+        except Exception as e:
+            logger.error("iRODS connection failed: %s", self.get_irods_error(e))
+            logger.error("Are you logged in? try 'iinit'")
             raise
-        files = dict(files=[], md5=[])
-        base_path = None
-        for line in ils_out.split("\n"):
-            m = re.fullmatch(r"(\s+)?(C-\s)?(\S+)\s*", line)
-            if m:
-                g = m.groups()
-                if not g[0]:
-                    base_path = g[2][:-1]
-                elif not g[1]:
-                    ftype = "files" if g[2][-4:] != ".md5" else "md5"
-                    files[ftype].append(os.path.join(base_path, g[2]))
-        return files
+
+    def get_data_objs(self, root_coll: iRODSCollection):
+        """Get data objects recursively under the given iRODS path."""
+        # NOTE: We shouldn't use walk() as it's extremely inefficient, but
+        #       calling SpecificQuery with a rodsuser results in SYS_NO_API_PRIV
```

Comment:
Can we ask on the irods support/mailing list/github issue about the `SYS_NO_API_PRIV` issue?

Reply:
I realized after pushing this commit that it's not exactly an issue: SpecificQuery basically allows writing any SQL to the iRODS database. I think there's a good reason why regular users are not allowed to do this. Wouldn't want a user to drop tables accidentally or maliciously. So walk() is, sadly, the way to go.